src/uu/rm/src/rm.rs

```diff
@@ -303,13 +303,31 @@ fn handle_dir(path: &Path, options: &Options) -> bool {
             }
         } else {
             let mut dirs: VecDeque<DirEntry> = VecDeque::new();
+            let mut not_descended: Vec<PathBuf> = Vec::new();
 
-            for entry in WalkDir::new(path) {
+            'outer: for entry in WalkDir::new(path) {
                 match entry {
                     Ok(entry) => {
+                        if options.interactive == InteractiveMode::Always {
+                            for not_descend in &not_descended {
+                                if entry.path().starts_with(not_descend) {
+                                    continue 'outer;
+                                }
+                            }
+                        }
                         let file_type = entry.file_type();
                         if file_type.is_dir() {
-                            dirs.push_back(entry);
+                            if options.interactive == InteractiveMode::Always
+                                && fs::read_dir(entry.path()).unwrap().count() != 0
+                            {
+                                if prompt_descend(entry.path()) {
+                                    dirs.push_back(entry);
+                                } else {
+                                    not_descended.push(entry.path().to_path_buf());
+                                }
+                            } else {
```

Comment:
please add a comment here. it is starting to be quite complex with the various if and else